src/java.base/share/classes/java/util/concurrent/CopyOnWriteArrayList.java

```diff
@@ -743,8 +743,7 @@ public boolean addAll(Collection<? extends E> c) {
             Object[] es = getArray();
             int len = es.length;
             Object[] newElements;
-            if (len == 0 && (c.getClass() == CopyOnWriteArrayList.class ||
-                             c.getClass() == ArrayList.class)) {
+            if (len == 0 && c.getClass().getModule() == Object.class.getModule()) {
```

Comment:
I think as soon as we anyway assign the array from trusted collection, we don't need `len == 0` check here any more

Reply:
> I think as soon as we anyway assign the array from trusted collection, we don't need `len == 0` check here any more

I think it is necessary to check the length here. 

If the length of the current COWArrayList is not zero, it is necessary to copy the elements to a larger new array.

Reply:
>  it is necessary to copy the elements to a larger new array

Right, I missed this is addAll() method. Btw, in this class you could do similar optimization in constructor as well.

Reply:
> > it is necessary to copy the elements to a larger new array
> 
> Right, I missed this is addAll() method. Btw, in this class you could do similar optimization in constructor as well.

Just now, I have pushed some updates and used similar optimizations in some places that were missed before.